llm/grammar.go

```diff
@@ -0,0 +1,537 @@
+				}
+
+				nextFourCharacters := classCharacters[i+2 : i+6]
+
+				if !isHexDigit(rune(nextFourCharacters[0])) || !isHexDigit(rune(nextFourCharacters[1])) || !isHexDigit(rune(nextFourCharacters[2])) || !isHexDigit(rune(nextFourCharacters[3])) {
+					return fmt.Errorf("invalid unicode escape character")
+				}
+
+				i += 6
+			} else {
+				i += 2
+			}
+		} else {
+			if classCharacters[i] == '^' {
+				return fmt.Errorf("^ should only be at the beginning of a character class")
+			}
+			i++
+		}
+	}
+
+	return nil
+}
+
+func validateStringLiteral(strLiteral string) error {
+	validEscapeCharacters := "\\\"ntrxu"
+
+	// make sure the string literal starts and ends with a quote
+	if len(strLiteral) < 2 || strLiteral[0] != '"' || strLiteral[len(strLiteral)-1] != '"' {
+		return fmt.Errorf("string literal must start and end with a quote")
+	}
+
+	i := 0
+	for i < len(strLiteral) {
+		if strLiteral[i] == '\\' {
+			// Check if the escape character is valid
+			if i+1 >= len(strLiteral) {
+				return fmt.Errorf("incomplete escape character")
+			}
+			escapeCharacter := strLiteral[i+1]
+			if !strings.Contains(validEscapeCharacters, string(escapeCharacter)) {
+				return fmt.Errorf("invalid escape character")
+			}
+			// if the escape character is 'x' or 'u', check if the following characters are hexadecimal digits
+			if escapeCharacter == 'x' {
+				// grab the following two characters
+				if i+3 >= len(strLiteral) {
+					return fmt.Errorf("incomplete hexadecimal escape character")
+				}
+
+				nextTwoCharacters := strLiteral[i+2 : i+4]
+
+				if !isHexDigit(rune(nextTwoCharacters[0])) || !isHexDigit(rune(nextTwoCharacters[1])) {
+					return fmt.Errorf("invalid hexadecimal escape character")
+				}
+
+				i += 4
+			} else if escapeCharacter == 'u' {
+				// grab the following four characters
+				if i+5 >= len(strLiteral) {
+					return fmt.Errorf("incomplete unicode escape character")
+				}
+
+				nextFourCharacters := strLiteral[i+2 : i+6]
+
+				if !isHexDigit(rune(nextFourCharacters[0])) || !isHexDigit(rune(nextFourCharacters[1])) || !isHexDigit(rune(nextFourCharacters[2])) || !isHexDigit(rune(nextFourCharacters[3])) {
+					return fmt.Errorf("invalid unicode escape character")
+				}
+
+				i += 6
+			} else {
+				i += 2
+			}
+		} else {
+			i++
+		}
+	}
+
+	return nil
+}
+
+func parseRule(rule string) ([]Token, error) {
+	var tokens []Token
+	rule = strings.TrimSpace(rule)
+	i := 0
+	n := len(rule)
+
+	for i < n {
+		switch {
+		case rule[i] == ' ':
+			// Skip spaces
+			i++
+		case rule[i] == '|':
+			tokens = append(tokens, Token{Type: TokenPipe, Value: string(rule[i])})
+			i++
+		case rule[i] == '(':
+			tokens = append(tokens, Token{Type: TokenLParen, Value: string(rule[i])})
+			i++
+		case rule[i] == ')':
+			tokens = append(tokens, Token{Type: TokenRParen, Value: string(rule[i])})
+			i++
+		case rule[i] == '[':
+			// Character class
+			start := i
+			i++
+			for i < n {
+				if rule[i] == ']' && rule[i-1] != '\\' {
```

Comment:
There is a defect on this line.

This defect can be triggered with the following BNF. The parser does not consider the `]` at the end of the first character class (`[^"\\]`) to be properly terminating, and therefore interprets everything up until the next `]` as part of this character class. So the parser produces the token "{ Type: TokenCharacterClass, Value `[^"\\] | "\\" (["\\/bfnrt]` }", which creates downstream issues in this case (it later errors with "unmatched closing parenthesis").
```bnf
string-char ::= [^"\\] | "\\" (["\\/bfnrt] | "u" [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F] [0-9a-fA-F])
```

I believe the six different scenarios are: `]`, `\]`, `\\]`, plus the same three scenarios at the beginning of the string. However, only `\]` should not terminate the character class. I believe, from my local testing, the correct condition is:

```go
if rule[i] == ']' && (i == 0 || rule[i-1] != '\\' || (i > 1 && rule[i-2] == '\\')) {
```